extension/src/Dictionary.ts

```diff
@@ -0,0 +1,165 @@
+import { existsSync, readFileSync, writeFileSync } from "fs";
+
+export class Dictionary implements IDictonary {
+  language: string;
+  wordList: DictPair[];
+  dictionaryFile: string;
+
+  constructor(dictionaryFile: string) {
+    this.dictionaryFile = dictionaryFile;
+    if (!existsSync(dictionaryFile)) {
+      throw new Error(`Could not open file: "${this.dictionaryFile}"`);
+    }
+    const d = JSON.parse(
+      readFileSync(this.dictionaryFile, "utf8")
+    ) as IDictonary;
+    this.language = d.language;
+    this.wordList = d.wordList;
+  }
+
+  toJSON(): unknown {
+    return {
+      language: this.language,
+      wordList: this.wordList,
+    };
+  }
+
+  saveDictionary(): void {
+    const spaces = 2;
+    writeFileSync(
+      this.dictionaryFile,
+      JSON.stringify(this.toJSON(), null, spaces),
+      "utf8"
+    );
+  }
+
+  addWord(word: string, replaceWith: string): Dictionary {
+    if (!this.exists(word)) {
+      const newWord: DictPair = {
+        word: word,
+        replacement: replaceWith,
+        settings: this.defaultSetting(),
+      };
+      this.wordList.push(newWord);
+      this.saveDictionary();
+    }
+    return this;
+  }
+
+  deleteWord(word: string): Dictionary {
+    this.wordList = this.wordList.filter((w) => w.word !== word);
+    this.saveDictionary();
+    return this;
+  }
+
+  find(word: string): DictPair | undefined {
+    return this.wordList.find(
+      (w) => w.word.toLocaleLowerCase() === word.toLocaleLowerCase()
+    );
+  }
+
+  searchAndReplace(text: string): string {
+    this.wordList.forEach((word) => {
+      word.settings = this.defaultSetting(word.settings);
+      const flags = word.settings.matchCasing ? "" : "i";
+      const re = new RegExp(`\\b${word.word}\\b`, flags);
```

Comment:
You are correct, this is a lapse in my reasoning. Need to work a bit on this :)

Reply:
I think I've got it covered now! :crossed_fingers: Take a close look in the tests for `Dictionary.replaceWord()`

I have a feeling that the usage of RegEx can be a bit tricky. In the case of `useRegEx = true` maybe `word.word` should be used as the single expression in `replaceWord`? It might also be a case that we don't have to support at the moment :)

Reply:
I thought that it had to be done in a loop, but that was much prettier. 👍